proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,150 @@
+
+## Introduction
+
+This proposal is to rework the design of the SDL core application manager layer in order to:
+ - Reduce coupling between components
+ - Reduce responsibility of the ApplicationManager class
+ - Provide a plugin architecture for adding additional functionality and RPC groups easily.
+
+## Motivation
+
+The SDL core project is actively implementing new features and new functionality.  
+To keep the project maintainable, on track and to reduce risks of regression during implementation
+refactoring of application manager layer is necessary.  
+Currently, extending SDL functionality may require code duplication. 
+
+Currently, extending SDL functionality may require code duplication. 
+The current RC module duplicates logic of the application_manager layer -- there is separate request controller, separate mechanism to call policy checks, and RPC's processing. Currently any common logic that is implemented in the application_manager layer needs to be duplicated in request controller -- this results in bugs and redundant code; it also requires more time for implementation.
+
+The following proposed solution will provide for better management and extensibility of SDL core RC components, without duplicating SDL core code:
+
+## Proposed solution
+#### Reduce coupling between components and use a dependency injection practice
+
+During component construction all dependencies should be explicitly received.
+This practice will improve encapsulation for the project and make code much more maintainable going forward.
+
+
+#### Reduce responsibility of the ApplicationManager class 
+Currently the application manager class has too many responsibilities:
+ - Service location
+ - Mobile/HMI RPC processing 
+ - Mobile/HMI RPC sending 
+ - Registration and managing mobile applications
+ - Handle different states of SDL (low voltage/ video streaming / audio_path_throw ...) 
+ 
+The purpose of this refactoring is to extract handling & sending RPC's to a separate module.
+ For example, many SDL commands require the application manager only for sending RPC's; these commands do not need to know information related to other application manager responsibilities.
+ Also, extracting RPC processing to separate modules will reduce regression risks related to changing functionalities in the ApplicationManager.
+ 
+#### Provide plugin architecture for adding additional functionality and RPC groups easily.
+
+SDL RPC's can be split into groups of RPC's related to certain functionality (for example RC or video streaming).  
+These RPC's may work with functionality that is not required by other RPC's, so exposure should be limited to the RPC's group.
+
+Adding new RPC or groups of RPC's is currently a very complicated and unclear process.
+
+The best solution is to adopt a plugin approach of adding new RPC's to SDL core.
+This will allow OEM's to implement code for new RPC's with ease by implement certain interfaces and provide them to SDL as shared library.
```

Comment:
This will allow OEM's to implement code for new RPC's with ease by implement certain interfaces and provide them to SDL as shared library.
should be
This will allow OEM's to implement code for new RPC's with ease by implementing certain interfaces and providing them to SDL as shared library.

Reply:
@mrapitis done in 850e6094ab6b9f6f3af20f2d82dd544c74d8265c